server/memstore.go

```diff
@@ -191,6 +192,25 @@ func (ms *memStore) storeRawMsg(subj string, hdr, msg []byte, seq uint64, ts int
 		ms.startAgeChk()
 	}
 
+	// In case this is a replica newer than the original one,
+	// wee need to consider that messages should be expired
+	// sooner than stream MaxAge value, as they might have
+	// older timestamps.
+	// This could be simpler, but we try to avoid calling time.Now() and other
+	// functions from time package, as this is a hot hpath.
+	// Especially that it should be called only once per scaling replicas.
+	if ms.cfg.MaxAge != 0 && ts > 0 && ts+int64(ms.cfg.MaxAge) < ms.ageChkTime {
+		now := time.Now().UnixNano()
+		// calculate duration when the next expireMsgs should be called.
+		fireIn := ts - (now - int64(ms.cfg.MaxAge))
+		// if previous age check was meant to happen later than the new one,
+		// update it.
+		if int64(ms.ageChkTime) >= fireIn+now {
+			ms.ageChkTime = now + fireIn
+			ms.ageChk = time.AfterFunc(time.Duration(fireIn), ms.expireMsgs)
```

Comment:
any chance that ageChk is being set twice in this codepath (first in `startAgeChk`?) just checking whether is needed to call cancelAgeChk so that previous timer does not fire

Reply:
It doesn't matter - as if not, it was initialized previously.
Switched to calling `ms.resetAgeChk`.

Could inline the if condition from reset, but as if we're here, it happens really rarely, I would prefer simplicity.

good catch, thx!